python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -108,46 +101,66 @@ def _convert_arrow_table_to_examples(
 
 def _get_feature_value(
     feature: "tf.train.Feature",
-) -> Union[List[bytes], List[float], List[int]]:
+) -> "pyarrow.Array":
+    import pyarrow as pa
+
     values = (
-        feature.bytes_list.value,
-        feature.float_list.value,
-        feature.int64_list.value,
+        feature.HasField("int64_list"),
+        feature.HasField("float_list"),
+        feature.HasField("bytes_list"),
     )
     # Exactly one of `bytes_list`, `float_list`, and `int64_list` should contain data.
+    # Otherwise, all three should be empty lists, indicating an empty feature value.
     assert sum(bool(value) for value in values) == 1
 
-    if feature.bytes_list.value:
-        return list(feature.bytes_list.value)
-    if feature.float_list.value:
-        return list(feature.float_list.value)
-    if feature.int64_list.value:
-        return list(feature.int64_list.value)
+    if feature.HasField("bytes_list"):
+        return pa.array(feature.bytes_list.value, type=pa.binary())
+    if feature.HasField("float_list"):
+        return pa.array(feature.float_list.value, type=pa.float32())
+    if feature.HasField("int64_list"):
+        return pa.array(feature.int64_list.value, type=pa.int64())
+    raise AssertionError(
+        "tf.train.Feature should have at least one value field set, "
+        "so this should never be reached."
+    )
```

Comment:
Here's an example tweak to pull some of the logic from `_convert_example_to_dict` into this function:
```suggestion
    if feature.HasField("bytes_list"):
        value = feature.bytes_list.value
        type_ = pa.binary()
    elif feature.HasField("float_list"):
        value = feature.float_list.value
        type_ = pa.float32()
    elif feature.HasField("int64_list"):
        value = feature.int64_list.value
        type_ = pa.int64()
    else:
        raise AssertionError(
            "tf.train.Feature should have at least one value field set, "
            "so this should never be reached."
        )
    if len(value) == 1:
        value = value[0]
    else:
        type_ = pa.list_(type_)
    return pa.array([value], type=type_)
```